modules/dlm/src/main/java/org/elasticsearch/dlm/DataLifecycleService.java

```diff
@@ -0,0 +1,321 @@
+
+                    long nowMillis = nowSupplier.getAsLong();
+                    if (nowMillis >= indexLifecycleDate.getMillis() + retention.getMillis()) {
+                        // there's an opportunity here to batch the delete requests (i.e. delete 100 indices / request)
+                        // let's start simple and reevaluate
+                        DeleteIndexRequest deleteRequest = new DeleteIndexRequest(backingIndex.getIndex().getName()).masterNodeTimeout(
+                            TimeValue.MAX_VALUE
+                        );
+
+                        // time to delete the index
+                        transportActionsDeduplicator.executeOnce(
+                            deleteRequest,
+                            ActionListener.noop(),
+                            (req, reqListener) -> deleteIndex(deleteRequest, reqListener)
+                        );
+                    }
+                }
+            }
+        }
+    }
+
+    private void rolloverDataStream(RolloverRequest rolloverRequest, ActionListener<Void> listener) {
+        // "saving" the rollover target name here so we don't capture the entire request
+        String rolloverTarget = rolloverRequest.getRolloverTarget();
+        client.admin().indices().rolloverIndex(rolloverRequest, new ActionListener<>() {
+            @Override
+            public void onResponse(RolloverResponse rolloverResponse) {
+                listener.onResponse(null);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                logger.error(() -> format("DLM rollover of [%s] failed", rolloverTarget), e);
+                listener.onFailure(e);
+            }
+        });
+    }
+
+    private void deleteIndex(DeleteIndexRequest deleteIndexRequest, ActionListener<Void> listener) {
+        assert deleteIndexRequest.indices() != null && deleteIndexRequest.indices().length == 1 : "DLM deletes one index at a time";
+        // "saving" the index name here so we don't capture the entire request
+        String targetIndex = deleteIndexRequest.indices()[0];
+        client.admin().indices().delete(deleteIndexRequest, new ActionListener<>() {
+            @Override
+            public void onResponse(AcknowledgedResponse acknowledgedResponse) {
+                listener.onResponse(null);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                // we're capturin
+                logger.error(() -> format("DLM request to delete [%s] failed", targetIndex), e);
+                listener.onFailure(e);
+            }
+        });
+    }
+
+    @Nullable
+    static TimeValue getRetentionConfiguration(DataStream dataStream) {
+        TimeValue retention = null;
+        if (dataStream.getLifecycle() != null) {
+            retention = dataStream.getLifecycle().getDataRetention();
+        }
+        return retention;
```

Comment:
Seems like this could be:

```suggestion
        if (dataStream.getLifecycle() == null) {
            return null;
        } else {
            return dataStream.getLifecycle().getDataRetention();
        }
```